src/actions/salesforce/campaigns/campaigns_send_data.ts

```diff
@@ -0,0 +1,143 @@
+import * as jsforce from "jsforce"
+import * as winston from "winston"
+import * as Hub from "../../../hub"
+import {CHUNK_SIZE, MemberIds, Tokens} from "../campaigns/salesforce_campaigns"
+import { sfdcConnFromRequest } from "../common/oauth_helper"
+
+interface CampaignMember {
+  ContactId?: string
+  LeadId?: string
+  CampaignId: string
+  Status?: string
+}
+
+interface MemberErrors {
+  memberId: string
+  errors: jsforce.ErrorResult
+}
+
+type MemberType = "ContactId" | "LeadId"
+
+export class SalesforceCampaignsSendData {
+  readonly oauthCreds: { oauthClientId: string; oauthClientSecret: string }
+
+  constructor(oauthClientId: string, oauthClientSecret: string) {
+    this.oauthCreds = { oauthClientId, oauthClientSecret }
+  }
+
+  async sendData(
+    request: Hub.ActionRequest,
+    memberIds: MemberIds[],
+    tokens: Tokens,
+  ) {
+    // const sfdcConn = await salesforceLogin(request);
+    const sfdcConn = await sfdcConnFromRequest(request, tokens, this.oauthCreds)
```

Comment:
It looks like this block is repeated from above. Likely worth moving into its own function.

Reply:
Turns out that listening for the on refresh event is unnecessary and I can just pass back the connection object which handles the refresh automatically. I've moved this logic to very end of both functions as the refresh event can occur at any time.